### DISCUSSION TITLE
Feature request: CROSS JOIN UNNEST(map_entries(col)) with alias for keys and values.

### DISCUSSION THREAD

**Discussion by fpixl** [duckdb/duckdb#8524] [Ideas]
When doing a cross join unnest with a map column, we cannot specify the alias for keys and values.
See the following example:

```SQL
SELECT id, stacked_col.k, stacked_col.v
FROM  table_name
CROSS JOIN UNNEST(map_entries(col)) AS t(stacked_col)
```

This will create a column called `stacked_col` which is a struct, eg. `struct(k varchar, v varchar)`.

My ask is to have aliases for keys and values similar to what we can do in Trino:

```SQL
SELECT id, stacked_col_keys, stacked_col_values
FROM  table_name
CROSS JOIN UNNEST(map_entries(col)) AS t(stacked_col_keys, stacked_col_values)
```

Thank you!

**Comment by Tishj** [comment]
Maps are lists of structs internally, map_entries() returns the internal representation, once we get to unnest we've already lost this context probably.

But it might be interesting to support aliasing struct column keys in general
i.e: `AS t( col(col_key, col_value))`

**Reply by fpixl** [reply]
Got it. Aliasing struct column keys would be amazing! I'm using maps as a work around so I can cross join. If we could do cross join with structs it would be way more efficient for our usecase.

So let me see if I got what you meant. Say we have `col <struct A: varchar, B: varchar, C varchar>`. You are suggesting supporting alias and expand to 3 columns? Eg:

```SQL
SELECT id, A, B C
FROM  table_name
CROSS JOIN UNNEST(col) AS t(col(A), col(B), col(C))
```

**Reply by Tishj** [reply]
Not exactly, I was thinking about aliasing the struct field names

I think we can already do what you want to do:
```sql
D select * from unnest(map_entries(map {'a' : 42, 'b' : 1337})::STRUCT(key VARCHAR, value INTEGER)[]) as tbl(a);
┌────────────────────────────────────────┐
│                   a                    │
│ struct("key" varchar, "value" integer) │
├────────────────────────────────────────┤
│ {'key': a, 'value': 42}                │
│ {'key': b, 'value': 1337}              │
└────────────────────────────────────────┘
```

```sql
D select a.* from unnest(map_entries(map {'a' : 42, 'b' : 1337})::STRUCT(key VARCHAR, value INTEGER)[]) as tbl(a);
┌─────────┬───────┐
│   key   │ value │
│ varchar │ int32 │
├─────────┼───────┤
│ a       │    42 │
│ b       │  1337 │
└─────────┴───────┘
```

same outcome:
```sql
D select * from (select a.* from unnest(map_entries(map {'a' : 42, 'b' : 1337})) as tbl(a)) t(key, value);
┌─────────┬───────┐
│   key   │ value │
│ varchar │ int32 │
├─────────┼───────┤
│ a       │    42 │
│ b       │  1337 │
└─────────┴───────┘
```

Though the syntax is slightly more terse

**Reply by Tishj** [reply]
I was thinking about enabling this syntax:
```sql
D select a.* from unnest(map_entries(map {'a' : 42, 'b' : 1337})) as tbl(a(key, value));
```

currently results in:
```
Error: Parser Error: syntax error at or near "("
LINE 1: ...s(map {'a' : 42, 'b' : 1337})) as tbl(a(key, value));
```

**Reply by fpixl** [reply]
I'm having a hard time replicating that query with my example. Can you help me out?

Say we have a table with this schema:
```
id: int32
col: struct<key1: varchar, key2: varchar, key3: varchar>. # struct, NOT list<struct>
```

Here's sample data:
```
id    col
1     {key1: "a", key2: "b", key3: "c"}
2     {key1: null, key2: null, key3: "d"}
```

How do I turn these structs into key value structs?

Ideally we get structs into key value columns so it's easy to query. Eg.
```
id   col.key   col.value
1    key1      a
1    key2      b
1    key3      c
2    key1      null
2    key2      null
2    key3      d
```

**Reply by fpixl** [reply]
> I was thinking about enabling this syntax:
> 
> D select a.* from unnest(map_entries(map {'a' : 42, 'b' : 1337})) as tbl(a(key, value));

Got it, this would be great too.

**Reply by Tishj** [reply]
I think UNNEST on the struct, combined with PIVOT could do what you want to do

Though I wouldn't immediately know the right query
